staging/src/k8s.io/component-helpers/scheduling/corev1/helpers.go

```diff
@@ -30,3 +34,102 @@ func PodPriority(pod *v1.Pod) int32 {
 	// name of the pod was empty. So, we resolve to the static default priority.
 	return 0
 }
+
+// NodeSelectorRequirementsAsSelector converts the []NodeSelectorRequirement api type into a struct that implements
+// labels.Selector.
+func NodeSelectorRequirementsAsSelector(nsm []v1.NodeSelectorRequirement) (labels.Selector, error) {
+	if len(nsm) == 0 {
+		return labels.Nothing(), nil
+	}
+	selector := labels.NewSelector()
+	for _, expr := range nsm {
+		var op selection.Operator
+		switch expr.Operator {
+		case v1.NodeSelectorOpIn:
+			op = selection.In
+		case v1.NodeSelectorOpNotIn:
+			op = selection.NotIn
+		case v1.NodeSelectorOpExists:
+			op = selection.Exists
+		case v1.NodeSelectorOpDoesNotExist:
+			op = selection.DoesNotExist
+		case v1.NodeSelectorOpGt:
+			op = selection.GreaterThan
+		case v1.NodeSelectorOpLt:
+			op = selection.LessThan
+		default:
+			return nil, fmt.Errorf("%q is not a valid node selector operator", expr.Operator)
+		}
+		r, err := labels.NewRequirement(expr.Key, op, expr.Values)
+		if err != nil {
+			return nil, err
+		}
+		selector = selector.Add(*r)
+	}
+	return selector, nil
+}
+
+// MatchNodeSelectorTerms checks whether the node labels and fields match node selector terms in ORed;
+// nil or empty term matches no objects.
+func MatchNodeSelectorTerms(
+	nodeSelectorTerms []v1.NodeSelectorTerm,
+	nodeLabels labels.Set,
+	nodeFields fields.Set,
+) bool {
+	for _, req := range nodeSelectorTerms {
+		// nil or empty term selects no objects
+		if len(req.MatchExpressions) == 0 && len(req.MatchFields) == 0 {
+			continue
+		}
+
+		if len(req.MatchExpressions) != 0 {
+			labelSelector, err := NodeSelectorRequirementsAsSelector(req.MatchExpressions)
+			if err != nil || !labelSelector.Matches(nodeLabels) {
+				continue
+			}
+		}
+
+		if len(req.MatchFields) != 0 {
+			fieldSelector, err := NodeSelectorRequirementsAsFieldSelector(req.MatchFields)
+			if err != nil || !fieldSelector.Matches(nodeFields) {
+				continue
+			}
+		}
+
+		return true
+	}
+
+	return false
+}
+
+// NodeSelectorRequirementsAsFieldSelector converts the []NodeSelectorRequirement core type into a struct that implements
+// fields.Selector.
+func NodeSelectorRequirementsAsFieldSelector(nsm []v1.NodeSelectorRequirement) (fields.Selector, error) {
```

Comment:
Do we need this exported?

Reply:
From the PR description:

This moves NodeSelectorRequirementsAsSelector from pkg/apis/core/v1 to k8s.io/component-helpers/scheduling/corev1 to make it more importable by external components, such as the Descheduler (which currently duplicates this function to avoid a dependency on k8s.io/kubernetes)

Reply:
https://github.com/search?q=org%3Akubernetes-sigs+NodeSelectorRequirementsAsFieldSelector&type=Code
https://github.com/search?q=org%3Akubernetes+NodeSelectorRequirementsAsFieldSelector&type=Code

do not show anything outside of pkg/apis/core/v1/helper.

Reply:
I moved this function because it's called in [`MatchNodeSelectorTerms`](https://github.com/kubernetes/kubernetes/pull/95531/files#diff-07e1f86867f973a6e35d248fdb1efd7f0d82cbf96976028806d5ef04c60cd704L317-L346) which calls both `NodeSelectorRequirementsAsFieldSelector` and `NodeSelectorRequirementsAsSelector` (the latter being the original goal of this PR).

Leaving `MatchNodeSelectorTerms` in `pkg/apis/core/v1` would have added another import (k8s.io/component-helpers) to `pkg/apis/core`, which requires an update to [import-restrictions.yaml](https://github.com/kubernetes/kubernetes/blob/943b1dbf53c405004f35090ccef0dcf3c8065b7a/staging/publishing/import-restrictions.yaml#L1-L21). Based on the comments there about adding new exceptions, I figured it was better to just move the related functions together.

Reply:
As far as it needing to be exported, probably not since like Jan said it's only used by this package